Cabal/Distribution/Simple/Program/Run.hs

```diff
@@ -243,13 +243,13 @@ multiStageProgramInvocation simple (initial, middle, final) args =
       chunkSize    = maxCommandLineSize - fixedArgSize
 
    in case splitChunks chunkSize args of
-        []     -> [ simple ]
+        []        -> [ simple ]
 
-        [c]    -> [ simple  `appendArgs` c ]
+        [c]       -> [ simple  `appendArgs` c ]
 
-        (c:cs) -> [ initial `appendArgs` c ]
-               ++ [ middle  `appendArgs` c'| c' <- init cs ]
-               ++ [ final   `appendArgs` c'| let c' = last cs ]
+        (c:c2:cs) -> [ initial `appendArgs` c ]
+                  ++ [ middle  `appendArgs` c'| c' <- init (c2:|cs) ]
```

Comment:
Don't we have `unsnoc :: NonEmpty a -> ([a], a)` ? feels boring to traverse the list twice (and mention it twice). Then we could written

```haskell
(c:c2:cs) | (xs, x) <- unsnoc (c2:|cs) -> [ ... ] ++ [ ... | c' <- xs ] ++ [ final `appendArgs` x ]
```

```haskell
unsnocNE :: NonEmpty a -> ([a], a)
unsnocNE (x:|xs) =  go x xs where
    go y []     = ([], y)
    go y (z:zs) = let ~(ws, w) = go z zs in (y : ws, w)
```

Reply:
Thanks!  What's the `~` called, so I can read up about it?

Reply:
irrefutable pattern, here it's not strictly needed as `let` bindings are irrefutable, but I added it to make it more explicit.

In general it affects lazyness, the strict pattern match on pair, e.g.

```haskell
go y (z:zs) = case go z zs of (ws, w) -> (y : ws, w)
```

would force the traversal of original list way to early. But not we can say:

```haskell
> take 20 $ fst $ uncons (0 :| [1..])
[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19]
```

(though, if we ask for the last value, that would diverge).